process/blockProcessor.go

```diff
@@ -1,26 +1,105 @@
+	withTxsParamTrue = "?withTxs=true"
+)
+
+type blockProcessor struct {
+	proc     Processor
 	dbReader ExternalStorageConnector
 }
 
 // NewBlockProcessor will create a new block processor
-func NewBlockProcessor(dbReader ExternalStorageConnector) (*dbBlockProcessor, error) {
+func NewBlockProcessor(dbReader ExternalStorageConnector, proc Processor) (*blockProcessor, error) {
 	if check.IfNil(dbReader) {
 		return nil, ErrNilDatabaseConnector
 	}
+	if check.IfNil(proc) {
+		return nil, ErrNilCoreProcessor
+	}
 
-	return &dbBlockProcessor{
+	return &blockProcessor{
 		dbReader: dbReader,
+		proc:     proc,
 	}, nil
 }
 
-// GetBlockByShardIDAndNonce return the block byte shardID and nonce
-func (bp *dbBlockProcessor) GetBlockByShardIDAndNonce(shardID uint32, nonce uint64) (data.ApiBlock, error) {
-	return bp.dbReader.GetBlockByShardIDAndNonce(shardID, nonce)
+// GetAtlasBlockByShardIDAndNonce return the block byte shardID and nonce
+func (bp *blockProcessor) GetAtlasBlockByShardIDAndNonce(shardID uint32, nonce uint64) (data.ApiBlock, error) {
+	return bp.dbReader.GetAtlasBlockByShardIDAndNonce(shardID, nonce)
+}
+
+// GetNetworkStatusMetrics will simply forward the network status metrics from an observer in the given shard
+func (bp *blockProcessor) GetBlockByHash(shardID uint32, hash string, withTxs bool) (*data.GenericAPIResponse, error) {
+	observers, err := bp.getObserversOrFullHistoryNodes(shardID)
+	if err != nil {
+		return nil, err
+	}
+
+	path := fmt.Sprintf("%s/%s", blockByHashPath, hash)
+	if withTxs {
+		path += withTxsParamTrue
+	}
+
+	for _, observer := range observers {
+		var response data.GenericAPIResponse
+
+		_, err := bp.proc.CallGetRestEndPoint(observer.Address, path, &response)
+		if err != nil {
+			log.Error("block request", "observer", observer.Address, "error", err.Error())
+			continue
+		}
+
+		log.Info("fblock request", "shard id", observer.ShardId, "hash", hash, "observer", observer.Address)
+		return &response, nil
+
+	}
+
+	return nil, ErrSendingRequest
+}
+
+func (bp *blockProcessor) GetBlockByNonce(shardID uint32, nonce uint64, withTxs bool) (*data.GenericAPIResponse, error) {
+	observers, err := bp.getObserversOrFullHistoryNodes(shardID)
+	if err != nil {
+		return nil, err
+	}
+
+	path := fmt.Sprintf("%s/%d", blockByNoncePath, nonce)
+	if withTxs {
+		path += withTxsParamTrue
+	}
+
+	for _, observer := range observers {
+		var response data.GenericAPIResponse
+
+		_, err := bp.proc.CallGetRestEndPoint(observer.Address, path, &response)
+		if err != nil {
+			log.Error("block request", "observer", observer.Address, "error", err.Error())
+			continue
+		}
+
+		log.Info("block request", "shard id", observer.ShardId, "nonce", nonce, "observer", observer.Address)
+		return &response, nil
+
+	}
+
+	return nil, ErrSendingRequest
+}
+
+func (bp *blockProcessor) getObserversOrFullHistoryNodes(shardID uint32) ([]*data.NodeData, error) {
+	fullHistoryNodes, err := bp.proc.GetFullHistoryNodes(shardID)
+	if err == nil {
```

Comment:
maybe you should change the logic err != nil

Reply:
no, it is more clear this way because this doesn't have to return a `nil` error. and if it does, we should return the full history nodes